src/lib.rs

```diff
@@ -0,0 +1,1362 @@
+    ))]
+    pub fn register_irqfd(&self, eventfd: RawFd, gsi: u32) -> Result<()> {
+        let irqfd = kvm_irqfd {
+            fd: eventfd as u32,
+            gsi,
+            ..Default::default()
+        };
+        // Safe because we know that our file is a VM fd, we know the kernel will only read the
+        // correct amount of memory from our pointer, and we verify the return result.
+        let ret = unsafe { ioctl_with_ref(self, KVM_IRQFD(), &irqfd) };
+        if ret == 0 {
+            Ok(())
+        } else {
+            Err(io::Error::last_os_error())
+        }
+    }
+
+    /// Constructs a new kvm VCPU fd.
+    ///
+    /// # Arguments
+    ///
+    /// * `id` - The CPU number between [0, max vcpus).
+    ///
+    /// # Errors
+    /// Returns an error when the VM fd is invalid or the VCPU memory cannot be mapped correctly.
+    ///
+    pub fn create_vcpu(&self, id: u8) -> Result<VcpuFd> {
+        // Safe because we know that vm is a VM fd and we verify the return result.
+        #[allow(clippy::cast_lossless)]
+        let vcpu_fd = unsafe { ioctl_with_val(&self.vm, KVM_CREATE_VCPU(), id as c_ulong) };
+        if vcpu_fd < 0 {
+            return Err(io::Error::last_os_error());
+        }
+
+        // Wrap the vcpu now in case the following ? returns early. This is safe because we verified
+        // the value of the fd and we own the fd.
+        let vcpu = unsafe { File::from_raw_fd(vcpu_fd) };
+
+        let kvm_run_ptr = KvmRunWrapper::from_fd(&vcpu, self.run_size)?;
+
+        Ok(VcpuFd { vcpu, kvm_run_ptr })
+    }
+}
+
+impl AsRawFd for VmFd {
+    fn as_raw_fd(&self) -> RawFd {
+        self.vm.as_raw_fd()
+    }
+}
+
+/// An address either in programmable I/O space or in memory mapped I/O space.
+pub enum IoEventAddress {
+    /// Representation of an programmable I/O address.
+    Pio(u64),
+    /// Representation of an memory mapped I/O address.
+    Mmio(u64),
+}
+
+/// Used in `VmFd::register_ioevent` to indicate that no datamatch is requested.
+pub struct NoDatamatch;
+impl Into<u64> for NoDatamatch {
+    fn into(self) -> u64 {
+        0
+    }
+}
+
+/// A safe wrapper over the `kvm_run` struct.
+///
+/// The wrapper is needed for sending the pointer to `kvm_run` between
+/// threads as raw pointers do not implement `Send` and `Sync`.
+pub struct KvmRunWrapper {
+    kvm_run_ptr: *mut u8,
+}
+
+// Send and Sync aren't automatically inherited for the raw address pointer.
+// Accessing that pointer is only done through the stateless interface which
+// allows the object to be shared by multiple threads without a decrease in
+// safety.
+unsafe impl Send for KvmRunWrapper {}
+unsafe impl Sync for KvmRunWrapper {}
+
+impl KvmRunWrapper {
+    /// Maps the first `size` bytes of the given `fd`.
+    ///
+    /// # Arguments
+    /// * `fd` - File descriptor to mmap from.
+    /// * `size` - Size of memory region in bytes.
+    pub fn from_fd(fd: &AsRawFd, size: usize) -> Result<KvmRunWrapper> {
```

Comment:
`mmap_from_fd` or simply `mmap` would be more explicit imho.

Reply:
Done.